docs/dev/guidelines/client-tests.rst

```diff
@@ -0,0 +1,257 @@
+
+    .. code:: ts
+
+        describe('ParticipationSubmissionComponent', () => {
+            ...
+
+            beforeEach(() => {
+                return TestBed.configureTestingModule({
+                    imports: [ArtemisTestModule, RouterTestingModule, NgxDatatableModule],
+                    declarations: [
+                        ParticipationSubmissionComponent,
+                        MockComponent(UpdatingResultComponent),
+                        MockComponent(AssessmentDetailComponent),
+                        MockComponent(ComplaintsForTutorComponent),
+                        MockTranslateValuesDirective,
+                        MockPipe(ArtemisTranslatePipe),
+                        MockPipe(ArtemisDatePipe),
+                        MockPipe(ArtemisTimeAgoPipe),
+                        MockDirective(DeleteButtonDirective),
+                        MockComponent(AlertComponent),
+                        MockComponent(ResultComponent),
+                        MockComponent(FaIconComponent),
+                    ],
+                    providers: [
+                        ...
+                    ],
+                })
+                    .overrideModule(ArtemisTestModule, { set: { declarations: [], exports: [] } })
+                    .compileComponents()
+                    .then(() => {
+                        ...
+                    });
+            });
+        });
+
+Running time: 13.685s:
+
+.. code-block:: text
+
+   PASS  src/test/javascript/spec/component/participation-submission/participation-submission.component.spec.ts (13.685 s, 535 MB heap size)
+
+Now the whole testing suite is running **~25 times faster**!
+
+Here are the improvements for the test above:
+
+* **Removed** production module imports:
+
+.. code-block:: text
+
+  - ArtemisResultModule
+  - ArtemisSharedModule
+  - TranslateModule.forRoot()
+  - MomentModule
+
+* **Mocked** pipes, directives and components that are not supposed to be tested:
+
+.. code-block:: text
+
+  + MockTranslateValuesDirective
+  + MockPipe(ArtemisTranslatePipe)
+  + MockPipe(ArtemisDatePipe)
+  + MockPipe(ArtemisTimeAgoPipe)
+  + MockDirective(DeleteButtonDirective)
+  + MockComponent(AlertComponent)
+  + MockComponent(ResultComponent)
+  + MockComponent(FaIconComponent)
+
+More examples on test speed improvement can be found in the following PR: https://github.com/ls1intum/Artemis/pull/3879/files
+
+Services should be mocked if they simply return some data from the server. However, if the service has some form of logic included (for exampling converting dates to moments),
+and this logic is important for the component, do not mock the service methods, but mock the http requests and responses from the api. This allows us to test the interaction
+of the component with the service and in addition test that the service logic works correctly. A good explanation can be found in the official angular documentation: https://angular.io/guide/http#testing-http-requests
```

Comment:
I am not sure we need to do it in cases when the service is just making an http request without any pre- and post-processing of the data. I guess otherwise we would be just testing the the right url with the right http method is used, right?

Reply:
yes kind of... but there exist a lot of `....service.spec.ts` files. However, I think the way the guideline is written here is valid too. Or this is a whole new topic that needs to be discussed and the guideline then adapted sometime later, if necessary.